admin/services/pupil-frozen.service/pupil-frozen.data.service.ts

```diff
@@ -0,0 +1,80 @@
+const sqlService = require('../data-access/sql.service')
+import { TYPES } from '../data-access/sql.service'
+
+export class PupilFrozenDataService {
+/**
+ * @description look up a pupils frozen flag by their id
+ * @param {number} pupilId
+ * @returns {Promise<Array<any>>}
+ */
+  public static async isFrozen (pupilId: number): Promise<any> {
+    const sql = 'SELECT frozen FROM mtc_admin.pupil WHERE id=@pupilId'
+    const params = [{
+      name: 'pupilId',
+      type: TYPES.Int,
+      value: pupilId
+    }]
+    return sqlService.readonlyQuery(sql, params)
+  }
+
+/**
+ * @description look up a pupils frozen flag by their urlSlug
+ * @param {string} pupilUrlSlug
+ * @returns {Promise<Array<any>>}
+ */
+  public static async isFrozenByUrlSlug (pupilUrlSlug: string): Promise<any> {
+    const sql = 'SELECT frozen FROM mtc_admin.pupil WHERE urlSlug=@pupilUrlSlug'
+    const params = [{
+      name: 'pupilUrlSlug',
+      type: TYPES.UniqueIdentifier,
+      value: pupilUrlSlug
+    }]
+    return sqlService.readonlyQuery(sql, params)
+  }
+
+/**
+ * @description get the count of pupils who are frozen within a set of urlSlugs
+ * @param {Array<string>} pupilUrlSlugs
+ * @returns {Promise<Array<any>>}
+ */
+  public static async countFrozenByUrlSlugs (pupilUrlSlugs: Array<string>): Promise<any> {
+    const { params, paramIdentifiers } = sqlService.buildParameterList(pupilUrlSlugs, TYPES.UniqueIdentifier,
+      'pupilUrlSlug')
+    const sql = `SELECT COUNT(id) as [frozenCount]
+      FROM mtc_admin.pupil WHERE frozen=1 AND urlSlug IN (${paramIdentifiers.join(', ')})`
+    return sqlService.readonlyQuery(sql, params)
+  }
+
+  /**
+ * @description get the count of pupils who are frozen within a set of urlSlugs
+ * @param {Array<string>} pupilIds
+ * @returns {Promise<Array<any>>}
+ */
+   public static async countFrozenByPupilIds (pupilIds: Array<number>): Promise<any> {
+    const paramData = sqlService.buildParameterList(pupilIds, TYPES.Int,
+      'pupilId')
+    const sql = `SELECT COUNT(id) as [frozenCount]
+      FROM mtc_admin.pupil WHERE frozen=1 AND id IN (${paramData.paramIdentifiers.join(', ')})`
+    return sqlService.readonlyQuery(sql, paramData.params)
+  }
+
+  public static async freezePupil (pupilUrlSlug: string): Promise<void> {
+    const sql = 'UPDATE mtc_admin.pupil SET frozen=1 WHERE urlSlug=@pupilUrlSlug'
+    const params = [{
+      name: 'pupilUrlSlug',
+      type: TYPES.UniqueIdentifier,
+      value: pupilUrlSlug
+    }]
+    return sqlService.modify(sql, params)
+  }
+
+  public static async thawPupil (pupilUrlSlug: string): Promise<void> {
```

Comment:
Docs missing
Returns something when should be void.

What happens why if we call thaw() on an already thawed pupil? Do we care enough to throw? Or emit an info log?

Reply:
i pondered this, and then got distracted.

lets discuss